### ISSUE TITLE
Propagate auth headers to white-listed resources

### ISSUE THREAD

**Issue by ikhomutov** [gogatekeeper/gatekeeper#534]
## Summary 
Right now if the user is accessing a white-listed resource, the authorization headers (neither Authorization nor X-Auth-Token) are forwarded to a upstream-url, which means that my backend is not aware if the user is authenticated or not. Previously there was an allow-anonymous attribute which did exactly what I need.
I could in theory use access token from cookies, but right now I have `enable-encrypted-token` parameter activated and I don't see easy ways to decrypt it.

## Why?
In my case, I have a website that provides an additional functionality for registered users, while others still can access the website in a read-only mode. I manage permissions on a backend side and need GGK to only handle authentication flow with keycloak.

## How
Probably it should be just like it was with allow-anonymous resource parameter

## Acceptance criteria
If the user navigates to a resource that is marked as white-listed, the authorization headers propagated to an upstream-url server, just like they do on protected resources.

**Comment by p53** [comment]
Hi thank you for opening issue! Hmm a little bit contradictory requirements, you would like to not authenticate as you are setting white listed resource on the other hand you would like to decrypt token that means you assume bounding of token to gatekeeper but other people might not want to do that. Not sure about allow anonymous attribute as far as I remember there was not such flag but yeah i don't know all deep history of louketo but sounds to me like an alternative to white-listed not additional flag. Anyway in old versions of louketo there were also bugs and bad design decisions so I need to think about this of how to implement it in clean way.

**Comment by ikhomutov** [comment]
yes, looks like you're right and there were no such flag. probably we used a patched version of ggk in my previous project and added this flag manually for this exact case.
still, I think this would be a very useful flag that allows to specify resources where you not force user to be authenticated, but if they are - you want to know this as it will change the appearance of the website.

**Comment by p53** [comment]
@ikhomutov i was thinking about this, actually you are already doing some kind of authentication as you must use password to encrypt token on client side, can you explain a little bit more your setup, you have client which sends encrypted token for both read-only and write users, you want gatekeeper to decrypt it (in case of write user to do also other checks) in both cases and then backend decides if user should have read-only or write access, right? wouldn't be proper design to authenticate both users but they would have different groups? why you don't want to authenticate read-only user?

**Comment by ikhomutov** [comment]
I think there is a misunderstanding. I use `enable-encrypted-token` parameter, which, I assumed, encrypt the value of access and refresh tokens in cookies on the GGK side. And my backend, where the requests are proxied to, cannot easily access them. If this is not how it works then my apologies for misleading.
My setup is just a simple combination of the GGK and Starlette based backend that renders pure html. And I need my backend to know if the user is authenticated or not even if the user requests publicly accessible pages like index page.
If you need a analogy of what I try to achieve - imagine youtube main page. If you are authorized, you have different side and nav bars than if you aren't, but the main content (list of videos) remains same. Right now if I specify my main page as whitelisted resource - the GGK will not propagate the authorization and other x-like headers to the upstream server thus I cannot render correct page for the specific user.
Hope I describe my issue understandably.

**Comment by p53** [comment]
Your analogy is not right, if you are anonymous on youtube you don't send any token. In your case you are assuming you are sending encrypted token, i don't understand why you want it to be whitelisted if you are sending token...that's quite not standard.

**Comment by ikhomutov** [comment]
Yes, if I am anonymous - I don't send any token. But if I am authorized - I should send token, even if I access publicly available page. Right now if I specify a resource as whitelisted, so this resource can be available to not authorized user, when authorized user access this resource the token is not sent to the backend. this is my issue.

**Comment by p53** [comment]
ok so actually what you want is:
- allow only NO token OR SPECIFIC token

**Comment by ikhomutov** [comment]
No. I want that if the request to any resource has access token in cookies - the GGK always include Authorization and/or X-Auth-Token headers to the request when forwarding to an upstream server. Right now it only does this for protected resources and doesn't do it for whitelisted resources

**Comment by p53** [comment]
so then you are sending token also to whitelisted resource as i already said before, why are you sending token to whitelisted resource?? It is against security to parse token => create headers from unverified token...

**Comment by ikhomutov** [comment]
Ok, I give up. Apparently, I don't completely understand how the ggk works and what it is used for. Sorry for bothering.

**Comment by hieunhh** [comment]
It looks like I also need a similar feature. In my case, I use Gogatekeeper to protect the documentation site (built with Astro Starlight framework in SSR mode), where some URLs are public, and some need to be restricted. All documentation pages have a Sidebar, and the list of URLs displayed on the sidebar is determined based on user authentication/authorization.

- For unauthenticated users, only public URLs are shown.
- For authenticated users with the appropriate permissions, both public and protected URLs are displayed.

Therefore, once a user has successfully authenticated and accesses a public URL, the upstream service must also receive authentication headers from Gogatekeeper to properly handle authorization and dynamically display the protected URLs that the user has permission to access in Sidebar.

Can Gogatekeeper support my issue?

**Comment by p53** [comment]
Hi @hieunhh, so actually you would like to have authentication on whitelisted resource if access cookie/bearer token is present (to create secure headers from token it must be verified/authenticated) and if it is not present then just let the people in without forwarded headers

**Comment by hieunhh** [comment]
Hi @p53 , thank you for your quick response. That is exactly what I am looking for—once a user has successfully authenticated, all URLs (both public and protected) should include user information in the request headers. However, since I am not yet fully familiar with the Gogatekeeper implementation, I am unsure whether this is feasible or if there are any potential performance issues or security vulnerabilities associated with it.

**Comment by p53** [comment]
@hieunhh ok thanks for response so we are in sync about idea 👍 , i will look on costs of implementing it

**Comment by p53** [comment]
@hieunhh you can try 3.3.0-rc2 `white-listed-anon`

**Comment by hieunhh** [comment]
@p53 Thank you so much for implementing this option. I really appreciate your quick support and effort. I'll test the implementation and will let you know how it goes.

**Comment by p53** [comment]
@hieunhh did you have any chance to test it, if it fits for you?
